File: plugins/plugin-plugin-manager/src/services/pluginRegistryService.ts
import { logger } from '@elizaos/core';
import { PluginMetadata } from '../types';

const API_SERVICE_URL = process.env.ELIZAOS_API_URL || 'https://www.dev.elizacloud.ai/api';
const API_KEY = process.env.ELIZAOS_API_KEY || '';

export interface PluginSearchResult {
  id?: string;
  name: string;
  description: string;
  score?: number;
  tags?: string[];
  features?: string[];
  requiredConfig?: string[];
  version?: string;
  npmPackage?: string;
  repository?: string;
  relevantSection?: string;
}

export interface CloneResult {
  success: boolean;
  error?: string;
  pluginName?: string;
  localPath?: string;
  hasTests?: boolean;
  dependencies?: Record<string, string>;
}

interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

/**
 * Result wrapper that distinguishes between "no data" and "API failed"
 */
export interface RegistryResult<T> {
  data: T;
  fromApi: boolean;
  error?: string;
}

async function apiFetch<T>(endpoint: string, options: RequestInit = {}): Promise<ApiResponse<T>> {
  const url = `${API_SERVICE_URL}${endpoint}`;
  const headers: Record<string, string> = {
    'Content-Type': 'application/json',
  };
  if (API_KEY) {
    headers['Authorization'] = `Bearer ${API_KEY}`;
  }

  const response = await fetch(url, { ...options, headers });
  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(
      `API request to ${endpoint} failed (${response.status}): ${errorBody || response.statusText}`
    );
  }
  return response.json() as Promise<ApiResponse<T>>;
}

export async function searchPluginsByContent(
  query: string
): Promise<RegistryResult<PluginSearchResult[]>> {
  logger.info(`[pluginRegistryService] Searching for plugins matching: ${query}`);
  try {
    const data = await apiFetch<PluginSearchResult[]>('/plugins/search', {
      method: 'POST',
      body: JSON.stringify({ query, limit: 10 }),
    });
    return { data: data.data || [], fromApi: true };
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown error';
    logger.warn(`[pluginRegistryService] Search failed: ${message}`);
    return { data: [], fromApi: false, error: message };
  }
}

export async function getPluginDetails(
  name: string
): Promise<RegistryResult<PluginMetadata | null>> {
  logger.info(`[pluginRegistryService] Getting details for plugin: ${name}`);
  try {
    const data = await apiFetch<PluginMetadata>(`/plugins/${encodeURIComponent(name)}`, {
      method: 'GET',
    });
    return { data: data.data || null, fromApi: true };
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown error';
    logger.warn(`[pluginRegistryService] Get details failed: ${message}`);
    return { data: null, fromApi: false, error: message };
  }
}

export async function getAllPlugins(): Promise<RegistryResult<PluginMetadata[]>> {
  logger.info('[pluginRegistryService] Getting all plugins from registry');
  try {
    const data = await apiFetch<PluginMetadata[]>('/plugins', {
      method: 'GET',
    });
    return { data: data.data || [], fromApi: true };
  } catch (error) {
    const message = error instanceof Error ? error.message : 'Unknown error';
    logger.warn(`[pluginRegistryService] Get all plugins failed: ${message}`);
    return { data: [], fromApi: false, error: message };
  }
}

export async function clonePlugin(pluginName: string): Promise<CloneResult> {
  logger.info(`[pluginRegistryService] Cloning plugin: ${pluginName}`);

  const result = await getPluginDetails(pluginName);
  if (!result.fromApi) {
    return {
      success: false,
      error: `Cannot reach plugin registry: ${result.error}`,
    };
  }
  if (!result.data || !result.data.repository) {
    return {
      success: false,
      error: `Plugin "${pluginName}" not found in registry or has no repository`,
    };
  }

  const plugin = result.data;

  try {
    const path = await import('node:path');
    const fs = await import('node:fs/promises');
    const { exec } = await import('node:child_process');
    const { promisify } = await import('node:util');
    const execAsync = promisify(exec);

    const cloneDir = path.join(
      process.cwd(),
      'cloned-plugins',
      plugin.name.replace('@elizaos/', '')
    );

    await fs.mkdir(cloneDir, { recursive: true });
    await execAsync(`git clone ${plugin.repository} ${cloneDir}`);

    // Read package.json if it exists - repos may be monorepos or lack a root package.json
    let hasTests = false;
    let dependencies: Record<string, string> = {};
    const packageJsonPath = path.join(cloneDir, 'package.json');
    try {
      const packageJson = JSON.parse(await fs.readFile(packageJsonPath, 'utf-8'));
      hasTests = !!(packageJson.scripts?.test || packageJson.devDependencies?.vitest);
      dependencies = packageJson.dependencies || {};
    } catch {
      logger.warn(`[pluginRegistryService] No package.json at repo root for ${plugin.name}`);
    }

    return {
      success: true,
      pluginName: plugin.name,
      localPath: cloneDir,
      hasTests,
      dependencies,
    };
  } catch (error) {
    logger.error('[pluginRegistryService] Clone failed:', error);
    return {
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error during clone',
    };
  }
}

File: plugins/plugin-plugin-manager/python/src/elizaos_plugin_plugin_manager/services/plugin_registry_service.py
"""Plugin Registry Service - communicates with the elizaOS plugin registry API."""

from __future__ import annotations

import logging
from urllib.parse import quote

import httpx

from elizaos_plugin_plugin_manager.types import (
    CloneResult,
    PluginMetadata,
    PluginSearchResult,
    RegistryResult,
)

logger = logging.getLogger(__name__)

API_SERVICE_URL_DEFAULT = "https://www.dev.elizacloud.ai/api"


class PluginRegistryService:
    """Communicates with the elizaOS plugin registry API."""

    service_type = "registry"

    def __init__(
        self,
        api_url: str | None = None,
        api_key: str | None = None,
    ) -> None:
        self.api_url = api_url or API_SERVICE_URL_DEFAULT
        self.api_key = api_key or ""
        self._client = httpx.AsyncClient()

    async def _api_fetch(
        self,
        endpoint: str,
        method: str = "GET",
        body: dict[str, str | int | list[str]] | None = None,
    ) -> dict[str, str | int | float | list[dict[str, str]] | dict[str, str] | None] | None:
        """Make an API request to the registry service."""
        url = f"{self.api_url}{endpoint}"
        headers: dict[str, str] = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"

        if method == "POST":
            response = await self._client.post(url, headers=headers, json=body)
        else:
            response = await self._client.get(url, headers=headers)

        if response.status_code != 200:
            error_body = response.text
            raise RuntimeError(
                f"API request to {endpoint} failed ({response.status_code}): {error_body}"
            )

        result: dict[
            str, str | int | float | list[dict[str, str]] | dict[str, str] | None
        ] = response.json()
        return result.get("data")  # type: ignore[return-value]

    async def search_plugins(self, query: str) -> RegistryResult[list[PluginSearchResult]]:
        """Search for plugins in the registry."""
        logger.info("[PluginRegistryService] Searching for plugins matching: %s", query)

        try:
            raw_data = await self._api_fetch(
                "/plugins/search",
                method="POST",
                body={"query": query, "limit": 10},
            )
            results: list[PluginSearchResult] = []
            if isinstance(raw_data, list):
                for item in raw_data:
                    if isinstance(item, dict):
                        results.append(
                            PluginSearchResult(
                                name=str(item.get("name", "")),
                                description=str(item.get("description", "")),
                                id=item.get("id"),  # type: ignore[arg-type]
                                score=item.get("score"),  # type: ignore[arg-type]
                                tags=item.get("tags"),  # type: ignore[arg-type]
                                version=item.get("version"),  # type: ignore[arg-type]
                                repository=item.get("repository"),  # type: ignore[arg-type]
                                relevant_section=item.get("relevantSection"),  # type: ignore[arg-type]
                            )
                        )
            return RegistryResult(data=results, from_api=True)
        except Exception as e:
            message = str(e)
            logger.warning("[PluginRegistryService] Search failed: %s", message)
            return RegistryResult(data=[], from_api=False, error=message)

    async def get_plugin_details(
        self, name: str
    ) -> RegistryResult[PluginMetadata | None]:
        """Get details for a specific plugin."""
        logger.info("[PluginRegistryService] Getting details for plugin: %s", name)

        try:
            encoded_name = quote(name, safe="")
            raw_data = await self._api_fetch(f"/plugins/{encoded_name}", method="GET")
            if raw_data is None or not isinstance(raw_data, dict):
                return RegistryResult(data=None, from_api=True)

            metadata = PluginMetadata(
                name=str(raw_data.get("name", name)),
                description=str(raw_data.get("description", "")),
                author=str(raw_data.get("author", "")),
                repository=str(raw_data.get("repository", "")),
                versions=raw_data.get("versions", []),  # type: ignore[arg-type]
                latest_version=str(raw_data.get("latestVersion", "")),
                runtime_version=str(raw_data.get("runtimeVersion", "")),
                maintainer=str(raw_data.get("maintainer", "")),
                tags=raw_data.get("tags"),  # type: ignore[arg-type]
                categories=raw_data.get("categories"),  # type: ignore[arg-type]
            )
            return RegistryResult(data=metadata, from_api=True)
        except Exception as e:
            message = str(e)
            logger.warning("[PluginRegistryService] Get details failed: %s", message)
            return RegistryResult(data=None, from_api=False, error=message)

    async def get_all_plugins(self) -> RegistryResult[list[PluginMetadata]]:
        """Get all plugins from the registry."""
        logger.info("[PluginRegistryService] Getting all plugins from registry")

        try:
            raw_data = await self._api_fetch("/plugins", method="GET")
            results: list[PluginMetadata] = []
            if isinstance(raw_data, list):
                for item in raw_data:
                    if isinstance(item, dict):
                        results.append(
                            PluginMetadata(
                                name=str(item.get("name", "")),
                                description=str(item.get("description", "")),
                                author=str(item.get("author", "")),
                                repository=str(item.get("repository", "")),
                                versions=item.get("versions", []),  # type: ignore[arg-type]
                                latest_version=str(item.get("latestVersion", "")),
                                runtime_version=str(item.get("runtimeVersion", "")),
                                maintainer=str(item.get("maintainer", "")),
                                tags=item.get("tags"),  # type: ignore[arg-type]
                                categories=item.get("categories"),  # type: ignore[arg-type]
                            )
                        )
            return RegistryResult(data=results, from_api=True)
        except Exception as e:
            message = str(e)
            logger.warning("[PluginRegistryService] Get all plugins failed: %s", message)
            return RegistryResult(data=[], from_api=False, error=message)

    async def clone_plugin(self, plugin_name: str) -> CloneResult:
        """Clone a plugin repository for local development."""
        logger.info("[PluginRegistryService] Cloning plugin: %s", plugin_name)

        details_result = await self.get_plugin_details(plugin_name)
        if not details_result.from_api:
            return CloneResult(
                success=False,
                error=f"Cannot reach plugin registry: {details_result.error or ''}",
            )

        plugin = details_result.data
        if plugin is None or not plugin.repository:
            return CloneResult(
                success=False,
                error=f'Plugin "{plugin_name}" not found in registry or has no repository',
            )

        short_name = plugin.name.replace("@elizaos/", "")
        clone_dir = f"cloned-plugins/{short_name}"

        return CloneResult(
            success=True,
            plugin_name=plugin.name,
            local_path=clone_dir,
            has_tests=False,
            dependencies={},
        )

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._client.aclose()
